pkg/ccl/changefeedccl/sink_kafka_connection_test.go

```diff
@@ -0,0 +1,194 @@
+		},
+		// Sanity check kafka sasl parameters.
+		{
+			name:          "param sasl_enabled must be a bool",
+			uri:           "kafka://nope/?sasl_enabled=maybe",
+			expectedError: "param sasl_enabled must be a bool",
+		},
+		{
+			name:          "param sasl_handshake must be a bool",
+			uri:           "kafka://nope/?sasl_enabled=true&sasl_handshake=maybe",
+			expectedError: "param sasl_handshake must be a bool",
+		},
+		{
+			name:          "sasl_enabled must be enabled to configure SASL handshake behavior",
+			uri:           "kafka://nope/?sasl_handshake=false",
+			expectedError: "sasl_enabled must be enabled to configure SASL handshake behavior",
+		},
+		{
+			name:          "sasl_user must be provided when SASL is enabled",
+			uri:           "kafka://nope/?sasl_enabled=true",
+			expectedError: "sasl_user must be provided when SASL is enabled",
+		},
+		{
+			name:          "sasl_password must be provided when SASL is enabled",
+			uri:           "kafka://nope/?sasl_enabled=true&sasl_user=a",
+			expectedError: "sasl_password must be provided when SASL is enabled",
+		},
+		{
+			name:          "sasl_enabled must be enabled if a SASL user is provided",
+			uri:           "kafka://nope/?sasl_user=a",
+			expectedError: "sasl_enabled must be enabled if a SASL user is provided",
+		},
+		{
+			name:          "sasl_enabled must be enabled if a SASL password is provided",
+			uri:           "kafka://nope/?sasl_password=a",
+			expectedError: "sasl_enabled must be enabled if a SASL password is provided",
+		},
+		{
+			name:          "sasl_enabled must be enabled to configure SASL mechanism",
+			uri:           "kafka://nope/?sasl_mechanism=SCRAM-SHA-256",
+			expectedError: "sasl_enabled must be enabled to configure SASL mechanism",
+		},
+		{
+			name:          "param sasl_mechanism must be one of SCRAM-SHA-256, SCRAM-SHA-512, or PLAIN",
+			uri:           "kafka://nope/?sasl_enabled=true&sasl_mechanism=unsuppported",
+			expectedError: "param sasl_mechanism must be one of SCRAM-SHA-256, SCRAM-SHA-512, or PLAIN",
+		},
+	} {
+		t.Run(tc.name, func(t *testing.T) {
+			sqlDB.ExpectErr(
+				t, tc.expectedError,
+				fmt.Sprintf(`CREATE EXTERNAL CONNECTION '%s' AS '%s'`, tc.name, tc.uri),
+			)
+		})
+	}
+
+	t.Run("changefeed-with-well-formed-uri", func(t *testing.T) {
+		sqlDB.Exec(t, `CREATE EXTERNAL CONNECTION nope AS 'kafka://nope'`)
+		sqlDB.ExpectErr(t, `client has run out of available brokers`,
+			`CREATE CHANGEFEED FOR foo INTO 'external://nope'`)
+
+		sqlDB.Exec(t, `CREATE EXTERNAL CONNECTION "nope-with-params" AS 'kafka://nope/?tls_enabled=true&insecure_tls_skip_verify=true&topic_name=foo'`)
+		sqlDB.ExpectErr(t, `client has run out of available brokers`,
+			`CREATE CHANGEFEED FOR foo INTO 'external://nope-with-params'`)
+
+		sqlDB.ExpectErr(
+			t, `client has run out of available brokers`,
+			`CREATE CHANGEFEED FOR foo INTO 'external://nope/' WITH kafka_sink_config='{"Flush": {"Messages": 100, "Frequency": "1s"}}'`,
```

Comment:
i'm not sure how I feel about establishing changefeed and relyiing on the fact that "kafka://nope" doesn't resolve to anything... Probably a fair assumption.. but maybe it's not something you should really worry about?

Alternatively, you can inject something like "wrapsink" https://github.com/cockroachdb/cockroach/blob/master/pkg/ccl/changefeedccl/testfeed_test.go#L534-L534 so that your test sink can override e.g. Dial method.
That dial method could just examine that you "see" unwrapped kafka URI, and just return an error.

Reply:
Makes sense, I'm still working on this.

Reply:
Added a wrapper that asserts on the underlying sink type. Let me know what you think.